test/test_heartbeat.py

```diff
@@ -0,0 +1,107 @@
+import pytest
+import rclpy
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.node import Node
+from interfaces.msg import Heartbeat
+from common_package_py.common_node import CommonNode
+import time
+
+
+def test_heartbeat_rate():
+    rclpy.init()
+    common_node = CommonNode("heartbeat")
+    test_node = rclpy.create_node("test")
+    
+    heartbeat_period = 0.5
+    executor = SingleThreadedExecutor()
+
+    last_msg = Heartbeat()
+    last_msg.tick = 0
+    last_msg.time_stamp = test_node.get_clock().now().to_msg()
+
+    def heartbeat_callback(msg):
+        nonlocal last_msg
+        
+        # Calculate the difference in seconds and nanoseconds separately
+        diff_sec = test_node.get_clock().now().to_msg().sec - last_msg.time_stamp.sec
```

Comment:
Hier könnte man direkt ein Time Objekt aus der msg machen und die dann subtrahieren, um eine Duration zu erhalten.
Siehe auch hier:
* https://github.com/ros2/rclpy/blob/e181737f19d6b8cb335b0326a40ec2075267c555/rclpy/rclpy/time.py#L166 und
* https://github.com/ros2/rclpy/blob/e181737f19d6b8cb335b0326a40ec2075267c555/rclpy/rclpy/time.py#L98
Würdest so auch dire Race condition fixen, wenn beim ersten call sec bei n ist und nsec knapp unter 10^9ns und beim zweiten call ist sec bei n+1 und nsec knapp über 0. In dem Fall ist man also fast eine  Sekunde "in die Vergangenheit gereist"

Reply:
Das Problem mit der race condition könnte man prinzipiell auch anders lösen, ich hab es aber so umgeschrieben wie empfohlen